### PR TITLE
Paper: Python in Data Science Research and Education (Paffenroth)

### PR DESCRIPTION
**Pull request by rcpaffenroth**
Good evening,

  Please find my submission for the SciPy2015 conference in this pull request.  Note, the date in the copyright notice and the page headers is incorrect.  It refers to an earlier version of the conference I believe.   In the README.md file is was requested:

- Please do not modify any files outside of your paper directory.

I followed that request and did not fix the error above.

   Thank you very much,

      Randy Paffenroth

Associate Professor
Mathematical Sciences Department
Data Science Program
Worcester Polytechnic Institute

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/15

### CODE REVIEW

**Comment by katyhuff** [comment]
@rcpaffenroth Regarding the wrong year: I think you may have cloned the repo before I fixed that (a few weeks ago). You can see on the buildbot that everyone's pull requests (including yours) are building with the right year: http://nipy.bic.berkeley.edu:5000/ .

I suspect this will be fixed if you merge the most recent version of the branch. Let me know.

**Comment by josef-pkt** [comment]
Just a general question from an "innocent bystander" and statsmodels maintainer.

How can you summarize python for data science without even mentioning statsmodels ?
the title is not "python for machine learning"

**Comment by stefanv** [comment]
Well, to be fair, Josef, data science can mean a lot of different things for a lot of people :)

**Comment by rcpaffenroth** [comment]
Good morning Josef,

    Thanks a lot for you comment and question.  I am certainly aware of the fine work done in the Statsmodels module!  Unfortunately, I have to admit that I don’t happen to use Statsmodels extensively in either my teaching or my research (though it would make a welcome addition).  However, I certainly see that my monograph would be improved by mentioning and referencing Statsmodels, and I would be happy to make that inclusion.

   Thanks again,

       Randy

**Comment by josef-pkt** [comment]
Thanks Randy,

I understand that users, developers, teachers and researchers prefer some packages over others.

However, I think that summaries like this

https://github.com/scipy-conference/scipy_proceedings/pull/142/files#diff-eec5a0456df3f0af42407a2e69174163R665

should include the most popular packages in Data Science in general. And it specifically mentions "statistics" which in Python is mostly statsmodels and scipy.stats for general purpose use.  Related, I think seaborn (and matplotlib) is used much more in data science for visualization than Mayav.

**Comment by katyhuff** [comment]
@josef-pkt Just as a data point - I think seaborn has a little way to go before it outstrips mayavi use in my particular community (though I don't doubt it probably eventually will). In any case, I just mention it to point out that the ranking of plotting tools will vary wildly depending on who you ask. (Bokeh, for example, is a favorite among my data sciencey peers, but I suspect that's not representative of the whole wide world of pythonic data science.) Suffice it to say there are a lot of libraries. Since this paper has mentioned both matplotlib and mayavi, that takes care of two major players in my book.

**Comment by josef-pkt** [comment]
I only really care about the inclusion of statsmodels as the main related statistics package.

Otherwise, I only have a very selective view of Data Science from the outside and mostly based on blog posts and online course material, and discussion of R or Python or both for Data Science.
e.g. http://www.dataschool.io/teaching-data-science/ or this http://cs109.github.io/2014/

**Comment by rcpaffenroth** [comment]
Good morning,

   As per the discussion with regards to this paper, I have added several references to the text, including Statsmodels, Seaborn, and Bokeh.  I have also promoted an acknowledgement to a co-author (Prof. Xiangnon Kong) and have changed the text accordingly.  
  Do I need to make a new pull request or will you just pull the HEAD of my 2015 branch at the appropriate moment?

   Thanks!

      Randy

**Comment by katyhuff** [comment]
All pushes to your 2015 branch will update this pr. So, as more review
comments appear, keep doing what you are doing! Thanks!

**Comment by ankurankan** [comment]
@rcpaffenroth A nice overview for using python in data science.
I have just one suggestion that it would be good to mention somewhere that the code would work only on python 2 (as the print statements in the 3rd code block would throw errors on python 3).

Off Topic: You can give pgmpy (I am one of the core developers) a try for your work in graphical models.

**Comment by rcpaffenroth** [comment]
Good morning Ankur,

     Thank you very much for you kind words and I will be happy to make the change you suggest.  I will also be sure to give pgmpy a try!

    Thanks again,

        Randy

**Comment by Carreau** [comment]
I read that in the train yesterday, good read, I just spotted on McK10 reference that was not linked to the bibliography. Nothing else to say otherwise.

**Comment by rcpaffenroth** [comment]
Good afternoon Matthias,

    Thanks a lot for your kind words.  I will make the change you suggest.

    Thanks again,

        Randy

**Comment by Carreau** [comment]
@katyhuff @jaberg I think all the comments have been addressed.

**Comment by ankurankan** [comment]
@katyhuff @jaberg I also think this is ready.

**Comment by jaberg** [comment]
Great job in the review process, thanks from the proceedings chairs :)

**Comment by minrk** [comment]
I think others have done a good job reviewing more substantive content. I have some English/prose comments that I'll make in-line.

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 1]
```diff
@@ -0,0 +1,985 @@
...
+  experience.
+
+.. class:: keywords
+
+   data science, education, machine learning
+
+Introduction
+------------
+
+Data Science is a burgeoning field of study that lays at the
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "lays" should be "lies" (lay is transitive, lie is intransitive)

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 2]
```diff
@@ -0,0 +1,985 @@
...
+intersection of statistics, computer science, and numerous applied
+scientific domains.  As is common within such *interdisciplinary*
+domains of study, Data Science education, mentoring, and research
+draws ideas from, and is inspired by, several other domains such as
+the mathematical sciences, computer science, and various business and
+application domains.  Perhaps just as importantly, student who wish to
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "students"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 3]
```diff
@@ -0,0 +1,985 @@
...
+interest to Data Science.
+
+Worcester Polytechnic Institute (WPI) has recently (fall 2014) begun
+admitting students into its new Data Science Master's degree program
+and, as of spring 2015, has also initiated a Data Science
+Ph.D. program.  Even at this early stage, the program has been quite
+fortunate to receive many more applications from students then it can
+reasonably admit.  The authors have the pleasure of being faculty
+members in the program and have the honor of teaching a number of the
+courses on offer.  In addition, as a long time Python users (for one
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "as long time Python users"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 4]
```diff
@@ -0,0 +1,985 @@
...
+of them since 1997 in fact [Paf99]_), the authors were intrigued by
+the possibility of leveraging Python in the graduate Data Science
+curriculum and this monograph describes some of the experiences, both
+successes and challenges, gained from that effort.  Of course, it is
+much too early to make any comments on the sustained effect of using
+Python for graduate Data Science education, however, perhaps the
+gentle reader will find some value in the authors' experiences, even
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "gentle" is probably unnecessary, here.

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 5]
```diff
@@ -0,0 +1,985 @@
...
+at this early date.
+
+Of course, we are not the first to suggest Python's effectiveness in
+an education and research environment.  In fact, the Python
+programming language is quite popular in numerous problem domains and
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: You refer to Python as "Python programming language" here and "Python scripting language" above. I would be consistent.

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 6]
```diff
@@ -0,0 +1,985 @@
...
+Python has seen wide used in education, see e.g., [Mye07]_ and
+[Sta00]_.  In fact, it ranks quite highly in many surveys of
+programming language popularity [OGr14]_, it is seeing substantial
+growth within the Data Science community [Sin14]_, and is generally
+speaking quite easy to learn [Lut13]_.
+
+However, it is not our purpose here to focus on Python in the large,
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "in the large" -> "in general"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 7]
```diff
@@ -0,0 +1,985 @@
...
+and in which the authors and others have leveraged Python, is DS501
+|--| "Introduction to Data Science".  The idea of DS501 is to provide
+an introductory overview of the many fields that comprise Data Science
+and it is intended that DS501 be one of the first classes a new
+student takes when entering the program.  Second, one of the
+authors has also used Python to support MA542 |--| "Regression
+Analysis".  MA542 is a somewhat more advanced class that is a (core)
+elective in the Data Science program as well as being a class taken by
+many students who are seeking degrees in the Mathematical Sciences
+department.  Finally, the authors mentor a number of student's
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: students'

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 8]
```diff
@@ -0,0 +1,985 @@
...
+research projects both within the Data Science Program, the
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: you use 'both', but there appear to be three programs here.

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 9]
```diff
@@ -0,0 +1,985 @@
...
+Mathematical Sciences Department, and the Computer Science Department.
+Many of these research projects leverage Python, in various ways, and
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: no comma after Python

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 10]
```diff
@@ -0,0 +1,985 @@
...
+having access to a common code base allows the various student
+projects to build off of one another.
+
+Two key themes will permeate our discussion in the following sections.
+First, Python provides easy access to a vast array of libraries.  Even
+though Data Science education and research draws from many other
+domains, Python was always there with a library ready to support our
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: it might be nice to refer to "the Python community" or even "the scientific Python community" here, instead of Python itself providing the libraries.

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 11]
```diff
@@ -0,0 +1,985 @@
...
+Space does not allow for a fulsome description of the class content
+and, in any event, such a discussion would distract us from our focus
+on Python.  However, in the authors' view, one important feature of
+such a class is that the students should be able to *get "their hands
+dirty" playing with real data* both early and often.  Students can
+often find inspiration by seeing the ideas developed as part of the
+lectures being  put to use on problems of practical interest.
+
+With all of the above in mind, it was decided to have four
+interconnected *case studies* as one of the major learning activities
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: should this be "as the major learning activities..."?

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 12]
```diff
@@ -0,0 +1,985 @@
...
+the python-twitter library [Ptw15]_.
+
+Another key feature of the case studies in DS501 is that we chose to
+use IPython notebooks [Per07]_ both to provide the assignments to the
+students and to have the students submit their results.  Using IPython
+notebooks for both of these tasks provided a number of advantages.
+First and foremost, it let the instructors provide the students
+with skeleton implementations of their assignments and allowed the
+students to focus on their learning objectives.  Second, it provide a
+uniform and easy to use development environment for the student's
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: students'

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 13]
```diff
@@ -0,0 +1,985 @@
...
+decide precisely what the question of interest is. Second, the student
+is expected to make hypotheses or conjectures based upon the
+definitions they come up with.  For example, the student might
+conjecture that men's and women's rating for certain genres are highly
+correlated, while for other genres their ratings more independent.
+Finally, the students must try to either prove, or just as
+interestingly, disprove their conjectures based upon the data.  
+
+Diving a bit more deeply into some of the specific functionality that
+we leverage in Python, we note that pandas [McK10]_ is particular
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: particularly

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 14]
```diff
@@ -0,0 +1,985 @@
...
+interest can only be answered by appropriate cross referencing between
+these three files.  For example, all three files must be referenced to
+answer a question as seemingly straight forward as "how many action
+movies do men rate higher than 4?"  While perhaps not too troublesome
+for students who are adept programmers, the cross referencing between
+the files presents an unnecessary impediment to less proficient
+students and overcoming this sort of impediment does not support the
+learning goals for this assignment.
+
+Of course, a straight forward answer would be for the instructors to
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "straightforward"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 15]
```diff
@@ -0,0 +1,985 @@
...
+Case Study Three
+~~~~~~~~~~~~~~~~
+
+The third case study is substantially more challenging than the second
+case study, but builds on the foundations already laid down.  While
+case study two focused on analyzing *numerical* movie reviews, case study
+three focuses on detecting positive and negative reviews from raw text
+using natural language processing.
+
+In particular, is case study three, the class turns its attention to the
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "in case study three"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 16]
```diff
@@ -0,0 +1,985 @@
...
+there are many open research problems in this domain.  On the other
+hand, such problems are clearly of importance in many domains, and it
+is not at all difficult to get students interested in solving them.
+The question remains, how can students in their very first Data
+Science class be expected to approach such difficult and important
+problems, and still be able to make meaningful progress?  Of course,
+the answer is, again, Python.
+
+In particular, we base this case study on the excellent scikit-learn
+[Ped11]_ Python library.  The scikit-learn provides easy to use and
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: remove "The"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 17]
```diff
@@ -0,0 +1,985 @@
...
+supported by a language such as Python and the libraries it provides.
+
+
+Case Study Four
+~~~~~~~~~~~~~~~
+
+The final case study, and in some sense the capstone of the class,
+revolves around the Yelp Dataset Challenge
+http://www.yelp.com/dataset_challenge.  This case study involves a
+large data set with approximately 42,153 business, 252,898 users, and
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: "businesses"

File: papers/randy_paffenroth/00_paffenroth.rst [hunk 18]
```diff
@@ -0,0 +1,985 @@
...
+
+* Statistical modeling:  Statsmodels [StM15]_
+* Convex optimization:  cvxopt [Dah06]_, CVXPY [Dia14]_
+* Deep learning:  Theano [Ber11]_
+* Graphical models:  libpgm [Kar14]_, pgmpy [Pgm15]_
+* Scientific visualization:  Mayavi [Ram11]_, Matplotlib [Hun07]_, Bokeh [Bok15]_, Seaborn [Was14]_
+
+Accordingly, students who are trained in classes such as DS501 and
+MA542 can leverage that training to get a running start on their
+research subjects.  Perhaps, this is the single biggest advantage of
```
Comment (minrk) [inline papers/randy_paffenroth/00_paffenroth.rst]: no comma after Perhaps

**Comment by minrk** [comment]
I've made a few minor grammatical comments throughout, but :+1: to the paper in general.

**Comment by rcpaffenroth** [comment]
Thank you very much!   I have  made the suggested changes and pushed the results to github.

   Thanks again,

      Randy

________________________________
